src/VisualStudio/Core/Next/Remote/ServiceHubRemoteHostClient.cs

```diff
@@ -129,22 +136,80 @@ private void OnRpcDisconnected(object sender, JsonRpcDisconnectedEventArgs e)
 
                 try
                 {
-                    var descriptor = new ServiceDescriptor(serviceName) { HostGroup = hostGroup };
                     return await client.RequestServiceAsync(descriptor, cancellationToken).ConfigureAwait(false);
                 }
                 catch (RemoteInvocationException ex)
                 {
-                    // RequestServiceAsync should never fail unless service itself is actually broken.
-                    // right now, we know only 1 case where it can randomly fail. but there might be more cases so 
-                    // adding non fatal watson here.
-                    FatalError.ReportWithoutCrash(ex);
+                    // save info only if it failed with different issue than before.
+                    if (lastException?.Message != ex.Message)
+                    {
+                        // RequestServiceAsync should never fail unless service itself is actually broken.
+                        // So far, we catched multiple issues from this NFW. so we will keep this NFW.
+                        WatsonReporter.Report("RequestServiceAsync Failed", ex, ReportDetailInfo);
+
+                        lastException = ex;
+                    }
                 }
 
                 // wait for retry_delayInMS before next try
                 await Task.Delay(retry_delayInMS, cancellationToken).ConfigureAwait(false);
             }
 
-            return Contract.FailWithReturn<Stream>("Fail to get service. look FatalError.s_reportedException for more detail");
+            // crash right away to get better dump. otherwise, we will get dump from async exception
+            // which most likely lost all valuable data
+            FatalError.ReportUnlessCanceled(lastException);
+            GC.KeepAlive(lastException);
+
+            // unreachable
+            throw ExceptionUtilities.Unreachable;
+        }
+
+        private static int ReportDetailInfo(IFaultUtility faultUtility)
+        {
+            // 0 means send watson, otherwise, cancel watson
+            // we always send watson since dump itself can have valuable data
+            var exitCode = 0;
+
+            try
+            {
+                var logPath = Path.Combine(Path.GetTempPath(), "servicehub", "logs");
+                if (!Directory.Exists(logPath))
+                {
+                    return exitCode;
+                }
+
+                // attach all log files that are modified less than 1 day before.
+                var now = DateTime.UtcNow;
+                var oneDay = TimeSpan.FromDays(1);
+
+                foreach (var file in Directory.EnumerateFiles(logPath, "*.log"))
+                {
+                    var lastWrite = File.GetLastWriteTimeUtc(file);
+                    if (now - lastWrite > oneDay)
+                    {
+                        continue;
+                    }
+
+                    faultUtility.AddFile(file);
```

Comment:
observation: you dereference `faultUtility` without checking that you were handed a non-null value. Will that be a problem?